### PR TITLE
Fix scrollbar margins

### LINKED ISSUE #106918: Scrollbar margins are handled inconsistently
crossAxisMargin is applied to the thumb and consumed by the track. 
mainAxisMargin is applied to both the track and the thumb.

I think the mainAxisMargin should apply to the thumb and be consumed by the track like in the cross axis.

Fix incoming.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      scrollBehavior: MaterialScrollBehavior().copyWith(scrollbars: false),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(),
      body: RawScrollbar(
        mainAxisMargin: 5,
        crossAxisMargin: 5,
        trackVisibility: true,
        thumbVisibility: true,
        thickness: 20,
        thumbColor: Colors.red,
        trackColor: Colors.blue,
        trackBorderColor: Colors.orange,
        child: Column(
          children: [
            Center(
              child: SizedBox(
                height: 500,
                child: CustomScrollView(
                  primary: true,
                  slivers: [
                    SliverList(
                      delegate: SliverChildBuilderDelegate(
                      (c, i) => Container(height: 50, child: Center(child: Text('Item $i'))),
                      childCount: 50,
                    ),
                    )
                  ]
                ),
              ),
            ),
            Container(height: 20, color: Colors.green)
          ],
        ),
      ), 
    );
  }
}

```

The track color issue here is separate. (https://github.com/flutter/flutter/issues/106917)

Screen Shot 2022-06-30 at 7 11 49 PM

### PR DESCRIPTION
**Pull request by Piinks**
Fixes https://github.com/flutter/flutter/issues/106918

This makes it such that the margins of the main and cross axes are treated the same way.
The track paints the full extent, while the margins are only applied to the thumb.

This is part of a yak shave in the scrollbar, as part of another proposal I found some bugs and code rot. I am fixing these all individually so that the proposal is not muddied with bugs fixes and new features mixed together.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I updated tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -1342,15 +1351,19 @@ class RawScrollbar extends StatefulWidget {
   /// {@macro flutter.widgets.Scrollbar.scrollbarOrientation}
   final ScrollbarOrientation? scrollbarOrientation;
 
-  /// Distance from the scrollbar's start and end to the edge of the viewport
-  /// in logical pixels. It affects the amount of available paint area.
+  /// Distance from the scrollbar thumb's start and end to the edge of the
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1354]: Distance from the scrollbar thumb's start or end to the nearest edge of the viewport

**Review by xu-baolin — COMMENTED** [review]
Should `_trackExtent` in the `getTrackToScroll` be replaced with traversable extent?

**Comment by Piinks** [comment]
> Should _trackExtent in the getTrackToScroll be replaced with traversable extent?

Yes!

Good catch. This is exactly the kind of thing I am trying to make more consistent to avoid future bugs in the larger clean up PR. I'll fix it here.

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -568,15 +568,15 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
         1.0);
 
     // Scrollbar thumb respects margin, track paints full length.
-    final double traversibleTrackExtent = _trackExtent - (2 * mainAxisMargin);
+    final double traversableTrackExtent = _trackExtent - (2 * mainAxisMargin);
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/scrollbar.dart:571]: I'll wrap this up in a getter in the larger refactor that will follow this.

**Comment by xu-baolin** [comment]
There are also some edge cases to confirm:
1. Is the margin area interactive?
2. we should skip paint if `traversibleTrackExtent` <= 0;
3.
https://github.com/flutter/flutter/blob/90d3b9d0db6dcde2b896d541911fc4ca009ae953/packages/flutter/lib/src/widgets/scrollbar.dart#L560-L564

I don't quite understand why we need to subtract the value of padding here, what do you think?

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 3]
```diff
@@ -1342,15 +1352,20 @@ class RawScrollbar extends StatefulWidget {
   /// {@macro flutter.widgets.Scrollbar.scrollbarOrientation}
   final ScrollbarOrientation? scrollbarOrientation;
 
-  /// Distance from the scrollbar's start and end to the edge of the viewport
-  /// in logical pixels. It affects the amount of available paint area.
+  /// Distance from the scrollbar thumb's start or end to the nearest edge of
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1355]: `ScrollbarPainter.mainAxisMargin`'s docs can update too.

**Comment by Piinks** [comment]
> There are also some edge cases to confirm:

Thank you so much for identifying these! I am following up on them and will have something back tomorrow. 🙏

**Review by pdblasi-google — APPROVED** [review]
LGTM

**Comment by xu-baolin** [comment]
I'm still a little worried about the impact of this change. First, there is a area where the thumb cannot reach in the main axle, which will mislead that the scrollable has not rolled to the edge of the view. Second, if we support buttons at both ends in the future, this margin area will be even more strange.

**Comment by Piinks** [comment]
> First, there is an area where the thumb cannot reach in the main axle, which will mislead that the scrollable has not rolled to the edge of the view.

This is actually the default for the Cupertino scrollbar, it has default margins such that the scrollbar never reaches the end of the perceived track area.

> Second, if we support buttons at both ends in the future, this margin area will be even more strange.

That is likely something we would address then, since this is currently hypothetical, we can't really account for it right now.

**Comment by Piinks** [comment]
Agh the bot was too fast for me. @xu-baolin if you have string feelings about this and would like to discuss further, we can revert this.

**Comment by xu-baolin** [comment]
> This is actually the default for the Cupertino scrollbar, it has default margins such that the scrollbar never reaches the end of the perceived track area.
> 

 The Cupertino scrollbar doesn't have track color, right?

**Comment by xu-baolin** [comment]
Considering that we have a 2D bar with track color, we need to configure margin in one corner to avoid crossover. It doesn't seem to be possible now, right?

**Comment by Piinks** [comment]
The motivation for this change was the inconsistently in how margins were being handled. The cross axis margin applied to the thumb, and not the track, and so it seemed logical for the margin in the main axis to reflect the same behavior.

Would the padding sample you demonstrated above better handle the 2D case you mention? I don't know that the main axis margin makes sense from an api standpoint to pad the scrollbar.
 
Perhaps padding should be exposed on the scrollbar instead of only using the MediaQueryData above it. It would give the user more control in a way that makes sense. We can certainly do that in a follow up change along with the clean up.

Also - we've been looking into and official 2D scrolling solution that would support 2D scrollbars. It's not quite ready yet, but I am optimistic that in this case it will just work, and not require extra configuring by the user.

**Comment by xu-baolin** [comment]
> Perhaps padding should be exposed on the scrollbar instead of only using the MediaQueryData above it.

Totally agree.

**Comment by xu-baolin** [comment]
LGTM

**Pull request merged**